### ISSUE TITLE
Extracting resources fails for lots of files

### ISSUE THREAD

**Issue by bitplane** [npjg/nefile#2]
Hey, cool project, but I think you need more test data :)

### source files

I was messing about with this collection from the Internet archive:

https://archive.org/details/630-windows-3x-games

### extracting

Throw it in a dir and `unzip` it, then from my bash_history, I did this:

```bash
for f in *.ZIP; do mkdir $(echo $f | cut -d '.' -f 1); mv $f $(echo $f | cut -d '.' -f 1); done
for d in $(find -type d); do pushd $d ; unzip *.ZIP; popd; done
rm */*.ZIP
# do the same for .zip too I guess
```

And a hacky script to export the resources:

```python
import nefile
import glob
import traceback


binaries = glob.glob('*/*.EXE')


for binary in binaries:
    try:
        f = nefile.nefile.NE(binary)
        f.export_resources(binary.split('/')[-2])
    except Exception as e:
        print('nope:', binary, type(e).__name__, str(e))
        traceback.print_exc()
```

### results

it errors out 90% of the time:

```bash
$ find -name '*.EXE' | wc -l
744

$ python export.py > errors.txt
$ wc -l errors.txt
687

$ cat errors.txt | cut -d ' ' -f 3- | sort | uniq -c | sort -nr

    362 AttributeError 'ApplicationDefinedData' object has no attribute 'write_ico_file'
    145 TypeError Cursor.__init__() takes from 2 to 3 positional arguments but 4 were given
     86 AttributeError 'StringTable' object has no attribute 'export'
     43 AssertionError 
     30 UnicodeDecodeError 'ascii' codec can't decode byte 0x80 in position 2: ordinal not in range(128)
     14 ValueError seek out of range
      2 UnicodeDecodeError 'ascii' codec can't decode byte 0xa9 in position 0: ordinal not in range(128)
      1 UnicodeDecodeError 'ascii' codec can't decode byte 0xb0 in position 24: ordinal not in range(128)
      1 UnicodeDecodeError 'ascii' codec can't decode byte 0xae in position 5: ordinal not in range(128)
      1 UnicodeDecodeError 'ascii' codec can't decode byte 0xa9 in position 81: ordinal not in range(128)
      1 UnicodeDecodeError 'ascii' codec can't decode byte 0xa3 in position 1: ordinal not in range(128)
      1 UnicodeDecodeError 'ascii' codec can't decode byte 0x8b in position 0: ordinal not in range(128)
```

### causes

In order of appearance

1. ~~Something weird going on here, caused by latest update? Can't eyeball this, need a debugger.~~ - looks to be fixed in git
2. Needs an extra dummy parameter adding to `Cursor.__init__()`, but PIL still fails afterwards if I do that on my machine. I installed Pillow rather than PIL because PIL is a dead project. Need to use `Image.load` not `Image`? edit: the bitmap stuff was confusing me, couldn't figure it out. I made the Cursor a subclass of Bitmap and dumped them to disk, but they look truncated in `hexdump`. Dunno if it's the bitmap writer, none of those exes have bitmap resources.
3. ~~Missing `export` method in `resources/string.py`~~ - looks fixed in git
4. Probably not a bug, but consider raising `ValueError` instead of `assert` [here](https://github.com/npjg/nefile/blob/ce096d29526d5bbc60ab68e09d30c034592f4fb3/nefile/nefile.py#L116)
5. and the others are probably caused by EXEs that have ~~different character encodings. Consider ignoring conversion errors, figure out the codepage, or cheat and use `chardet` to detect the encoding before converting to a string.~~ - nope they're zero terminated, see #3 
6. Haven't looked into this. edit: had a little look, but couldn't figure it out yet

I'll raise a pull request or two :)

**Comment by npjg** [comment]
Thank you for investigating! Yeah, life has been a bit too crazy for me to put much time into improving robustness here :( I appreciate your investigation and PRs though - looking at those now :)

**Comment by npjg** [comment]
I am making some other fixes in the next few days that should hopefully get all these EXEs supported. The cursor handling was indeed very messed up.

**Comment by npjg** [comment]
Try now with the newly released version 0.9.2? Now I can export from EXEs and DLLs in that dataset 😁. There are a good number in that set that appear to not actually be NEs (either PEs or plain DOS). I will continue to investigate/validate those. Also there is just a bit of weirdness still going on with icon exporting, but the vast majority of resources seem to export correctly.
